games/rpg/src/lib.rs

```diff
@@ -0,0 +1,1483 @@
+                    0..=32 => 1,
+                    33..=65 => 2,
+                    66..=98 => 3,
+                    _ => 4,
+                };
+                if choice != 4 {
+                    return choice
+                }
+                else {
+                    continue
+                }
+            };
+        }
+        // Takes two of the same type + level Weapon/Armor/Accessory NFT, burns them, and makes a new one based off of the first. 
+        // Takes stats from first NFT. Stats increase by 20% per upgrade
+        pub fn fuse_items(&mut self, which: u8, item_bucket: Bucket) -> Bucket {
+            assert!(item_bucket.amount() == dec!("2"));
+            assert!(item_bucket.resource_address() == self.item_nft);
+            if which == 1 {
+                let mut item: Weapon = item_bucket.non_fungibles()[0].data();
+                let item2: Weapon = item_bucket.non_fungibles()[1].data();
+                assert!(item.level == item2.level);
+                assert!(item.name == item2.name);
+                item.level += dec!(1);
+                item.physical_base *= dec!("1.2");
+                item.physical_scaling *= dec!("1.2");
+                item.spell_base *= dec!("1.2");
+                item.spell_scaling *= dec!("1.2"); 
+                item.ability_scaling *= dec!("1.2");          
+                let new = Weapon { class: item.class, name: item.name, level: item.level, physical_base: item.physical_base,
+                    physical_scaling: item.physical_scaling, spell_base: item.spell_base, spell_scaling: item.spell_scaling, ability: item.ability,
+                    ability_scaling: item.ability_scaling, range: item.range, version: self.version,
+                };
+                self.system_vault.authorize(||item_bucket.burn());
+                let new_bucket = self.system_vault.authorize(||
+                    borrow_resource_manager!(self.item_nft)
+                        .mint_non_fungible(&NonFungibleId::random(), new));
+                return new_bucket
+            }
+            else if which == 2 {
+                let mut item: Armor = item_bucket.non_fungibles()[0].data();
+                let item2: Armor = item_bucket.non_fungibles()[1].data();   
+                assert!(item.level == item2.level);            
+                assert!(item.id == item2.id);
+                assert!(item.part == item2.part);
+                item.level += dec!(1); 
+                item.health *= dec!("1.2");
+                item.defense *= dec!("1.2");
+                item.ability_scaling *= dec!("1.2");  
+                let new = Armor { id: item.id, part: item.part, level: item.level, health: item.health,
+                    defense: item.defense, weight: item.weight, ability: item.ability, ability_scaling: item.ability_scaling, version: self.version,
+                    };
+                self.system_vault.authorize(||item_bucket.burn());
```

Comment:
I see that this method uses `system_vault.authorize()` quite a number of times. Should this method be made to use `ComponentAuthZone::push(system_vault.create_proof())` at the top and `ComponentAuthZone::pop(system_vault.create_proof())` at the bottom instead?